platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java

```diff
@@ -37,32 +37,37 @@
 import org.gradle.api.internal.tasks.TaskDependencyFactory;
 import org.gradle.api.provider.Provider;
 import org.gradle.internal.DisplayName;
-import org.gradle.internal.FinalizableValue;
 import org.gradle.internal.reflect.Instantiator;
 import org.gradle.internal.typeconversion.NotationParser;
 
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
+import java.util.function.Supplier;
 
-public class DefaultConfigurationPublications implements ConfigurationPublications, FinalizableValue {
+public class DefaultConfigurationPublications implements ConfigurationPublications {
+
+    // Parent state
     private final DisplayName displayName;
-    private final PublishArtifactSet artifacts;
     private final PublishArtifactSetProvider allArtifacts;
     private final AttributeContainerInternal parentAttributes;
-    private final AttributeContainerInternal attributes;
+
+    // Services
     private final Instantiator instantiator;
     private final NotationParser<Object, ConfigurablePublishArtifact> artifactNotationParser;
     private final NotationParser<Object, Capability> capabilityNotationParser;
     private final FileCollectionFactory fileCollectionFactory;
     private final AttributesFactory attributesFactory;
     private final DomainObjectCollectionFactory domainObjectCollectionFactory;
     private final TaskDependencyFactory taskDependencyFactory;
+
+    // Mutable state
+    private final PublishArtifactSet artifacts;
+    private final AttributeContainerInternal attributes;
     private NamedDomainObjectContainer<ConfigurationVariant> variants;
-    private ConfigurationVariantFactory variantFactory;
     private DomainObjectSet<Capability> capabilities;
-    private boolean canCreate = true;
+    private Supplier<String> observationReason;
```

Comment:
🤔 It's not your invention, but can this just be made a String instead of a supplier?  I don't think we're saving many cycles by deferring a couple of concatenations here, and it add a little complexity that isn't really needed.

Reply:
Its not a matter of cycles, but memory usage. This avoids each configuration creating a new long string and retaining it for the entire build. Storing _how_ to make that string, if necessary, saves memory.